File: JavaScript/Codility/MaxCounters.js
// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(N, A) {
    var counter = new Array(N);
    for (var i=0; i < N; i++){
        counter[i] = 0;
    }
    var maxCounter = 0;
    var lastMaxCounter = 0;

    for (var K=0; K < A.length; K++){
        if (A[K] === N+1){
           lastMaxCounter = maxCounter;
        }
        else {
            if (counter[A[K]-1] < lastMaxCounter){
                counter[A[K]-1] = lastMaxCounter;
            }
            
            counter[A[K]-1]++;
            
            if (counter[A[K]-1] > maxCounter){
                maxCounter = counter[A[K]-1];
            }
        }
    }
    
    for (var i=0; i < N; i++){
        if (counter[i] < lastMaxCounter){
            counter[i] = lastMaxCounter;
        }
    }

    return counter;
}

File: Python/Codility/MaxCounters.py
def solution(N, A):
    counter = [0] * N
    maxCounter = 0
    lastMaxCounter = 0

    for K in range(len(A)):
        if A[K] == N + 1:
            lastMaxCounter = maxCounter
        else:
            counter[A[K] - 1] = max(lastMaxCounter, counter[A[K] - 1])
            counter[A[K] - 1] += 1
            maxCounter = max(maxCounter, counter[A[K] - 1])

    for i in range(N):
        counter[i] = max(lastMaxCounter, counter[i])

    return counter
